libs/shared/shared_common.c

```diff
@@ -1,8 +1,151 @@
 // Copyright (c) Microsoft Corporation
 // SPDX-License-Identifier: MIT
 
+#include "ebpf_program_types.h"
 #include "ebpf_shared_framework.h"
 
+enum _extension_object_type
+{
+    EBPF_PROGRAM_TYPE_DESCRIPTOR = 0,
+    EBPF_HELPER_FUNCTION_PROTOTYPE,
+    EBPF_PROGRAM_INFO,
+    EBPF_PROGRAM_DATA,
+    EBPF_PROGRAM_SECTION,
+};
+
+// Supported version and sizes of the various extension data structures.
+
+uint16_t _supported_ebpf_extension_version[] = {
+    EBPF_PROGRAM_TYPE_DESCRIPTOR_CURRENT_VERSION,
+    EBPF_HELPER_FUNCTION_PROTOTYPE_CURRENT_VERSION,
+    EBPF_PROGRAM_INFORMATION_CURRENT_VERSION,
+    EBPF_PROGRAM_DATA_CURRENT_VERSION,
+    EBPF_PROGRAM_SECTION_INFORMATION_CURRENT_VERSION,
+};
+
+#define EBPF_PROGRAM_TYPE_DESCRIPTOR_SIZE_0 EBPF_OFFSET_OF(ebpf_program_type_descriptor_t, is_privileged) + sizeof(char)
+size_t _ebpf_program_type_descriptor_supported_size[] = {EBPF_PROGRAM_TYPE_DESCRIPTOR_SIZE_0};
+
+#define EBPF_HELPER_FUNCTION_PROTOTYPE_SIZE_0 \
+    EBPF_OFFSET_OF(ebpf_helper_function_prototype_t, arguments) + 5 * sizeof(ebpf_argument_type_t)
+size_t _ebpf_helper_function_prototype_supported_size[] = {EBPF_HELPER_FUNCTION_PROTOTYPE_SIZE_0};
```

Comment:
It seems we are defining the macro for structure sizes twice. For example,

```
#define EBPF_HELPER_FUNCTION_PROTOTYPE_SIZE_0 \
    EBPF_OFFSET_OF(ebpf_helper_function_prototype_t, arguments) + 5 * sizeof(ebpf_argument_type_t)
```
and
```
#define EBPF_HELPER_FUNCTION_PROTOTYPE_CURRENT_VERSION_SIZE \
    (EBPF_OFFSET_OF(ebpf_helper_function_prototype_t, arguments) + 5 * sizeof(ebpf_argument_type_t))
```

I think we can define them only once, and `EBPF_HELPER_FUNCTION_PROTOTYPE_CURRENT_VERSION_SIZE` can be defined to `EBPF_HELPER_FUNCTION_PROTOTYPE_SIZE_0`

Reply:
This is by design, because with non-breaking changes, we will update the second macro to have _SIZE_N+1 with a different offset.